File: theory_of_encoding_compression_and_information_recovery/course_work/src/compressor/domain/users/services/user_service.py
from typing import Final

from compressor.domain.common.services.base import DomainService
from compressor.domain.users.entities.user import User
from compressor.domain.users.errors import RoleAssignmentNotPermittedError, CantChangeUsernameError
from compressor.domain.users.ports.password_hasher import PasswordHasher
from compressor.domain.users.ports.user_id_generator import UserIDGenerator
from compressor.domain.users.values.user_id import UserID
from compressor.domain.users.values.user_password_hash import UserPasswordHash
from compressor.domain.users.values.user_raw_password import UserRawPassword
from compressor.domain.users.values.user_role import UserRole
from compressor.domain.users.values.username import Username


class UserService(DomainService):
    def __init__(
            self,
            user_id_generator: UserIDGenerator,
            password_hasher: PasswordHasher,
    ) -> None:
        super().__init__()
        self._user_id_generator: Final[UserIDGenerator] = user_id_generator
        self._password_hasher: Final[PasswordHasher] = password_hasher

    def create(
            self,
            username: Username,
            raw_password: UserRawPassword,
            role: UserRole = UserRole.USER,
            is_active: bool = True,
    ) -> User:
        if not role.is_assignable:
            msg: str = f"Assignment of role {role} is not permitted."
            raise RoleAssignmentNotPermittedError(msg)

        user_id: UserID = self._user_id_generator()
        password_hash: UserPasswordHash = self._password_hasher.hash(raw_password)

        return User(
            id=user_id,
            username=username,
            role=role,
            is_active=is_active,
            password_hash=password_hash,
        )

    def is_password_valid(self, raw_password: UserRawPassword, user: User) -> bool:
        return self._password_hasher.verify(
            raw_password=raw_password,
            hashed_password=user.password_hash,
        )

    def change_password(self, user: User, raw_password: UserRawPassword) -> None:
        hashed_password: UserPasswordHash = self._password_hasher.hash(raw_password)
        user.password_hash = hashed_password

    # noinspection PyMethodMayBeStatic
    def change_username(self, user: User, new_username: Username) -> None:
        if user.role == UserRole.SUPER_ADMIN:
            msg: str = f"User {user.username} is not allowed to change username."
            raise CantChangeUsernameError(msg)
        user.username = new_username

File: сross_platform_software_systems/14/user_service/user_service/src/main/java/com/c3equalz/user_service/domain/user/services/UserService.java
package com.c3equalz.user_service.domain.user.services;

import com.c3equalz.user_service.domain.common.services.BaseDomainService;
import com.c3equalz.user_service.domain.user.entities.User;
import com.c3equalz.user_service.domain.user.errors.RoleAssignmentNotPermittedError;
import com.c3equalz.user_service.domain.user.events.UserChangedEmailEvent;
import com.c3equalz.user_service.domain.user.events.UserChangedNameEvent;
import com.c3equalz.user_service.domain.user.events.UserChangedPasswordEvent;
import com.c3equalz.user_service.domain.user.events.UserCreatedEvent;
import com.c3equalz.user_service.domain.user.ports.PasswordHasher;
import com.c3equalz.user_service.domain.user.ports.UserIdGenerator;
import com.c3equalz.user_service.domain.user.values.*;

import java.time.Instant;

/**
 * Domain service for users.
 * Provides methods for user management operations such as creation, password changes, etc.
 */
public class UserService extends BaseDomainService {

    private final PasswordHasher passwordHasher;
    private final UserIdGenerator userIdGenerator;

    /**
     * Creates a new UserService.
     *
     * @param passwordHasher the password hasher service
     * @param userIdGenerator the user ID generator
     */
    public UserService(PasswordHasher passwordHasher, UserIdGenerator userIdGenerator) {
        this.passwordHasher = passwordHasher;
        this.userIdGenerator = userIdGenerator;
    }

    /**
     * Fabric method that creates a new user. Produces event that user was created
     *
     * @param email email of the user
     * @param name username of the user
     * @param rawPassword raw password of the user
     * @param role role of the user (defaults to USER if not specified)
     * @return User entity if all checks passed
     * @throws RoleAssignmentNotPermittedError if role assignment is not permitted
     */
    public User create(
            UserEmail email,
            Username name,
            RawPassword rawPassword,
            UserRole role
    ) {
        if (role == null) {
            role = UserRole.USER;
        }

        if (!role.isAssignable()) {
            String msg = String.format("Assignment of role: %s not permitted.", role);
            throw new RoleAssignmentNotPermittedError(msg);
        }

        UserPasswordHash hashedPassword = passwordHasher.hash(rawPassword);
        UserID newUserId = userIdGenerator.generate();

        User newUser = new User(
                newUserId,
                email,
                name,
                hashedPassword,
                role
        );

        UserCreatedEvent newEvent = new UserCreatedEvent(
                newUserId,
                email.getValue(),
                name.getValue(),
                role
        );

        recordEvent(newEvent);

        return newUser;
    }

    /**
     * Method that checks if the given password is valid for the given user.
     *
     * @param user user entity which contains password
     * @param rawPassword password to check
     * @return true if the password is valid, false otherwise
     */
    public boolean isPasswordValid(User user, RawPassword rawPassword) {
        return passwordHasher.verify(rawPassword, user.getHashedPassword());
    }

    /**
     * Method that changes the password of the given user. Produces event that user changed the password
     *
     * @param user user entity which contains password
     * @param rawPassword new password to change
     */
    public void changePassword(User user, RawPassword rawPassword) {
        UserPasswordHash hashedPassword = passwordHasher.hash(rawPassword);
        user.setHashedPassword(hashedPassword);
        user.setUpdatedAt(Instant.now());

        UserChangedPasswordEvent newEvent = new UserChangedPasswordEvent(
                user.getId(),
                user.getName().getValue(),
                user.getEmail().getValue(),
                user.getRole().toString()
        );

        recordEvent(newEvent);
    }

    /**
     * Method that changes the name of the given user. Produces event that user changed the name
     *
     * @param user existing user entity in a database
     * @param newUserName new username for entity
     */
    public void changeName(User user, Username newUserName) {
        Username oldUserName = user.getName();
        user.setName(newUserName);
        user.setUpdatedAt(Instant.now());

        UserChangedNameEvent newEvent = new UserChangedNameEvent(
                user.getId(),
                oldUserName.getValue(),
                newUserName.getValue(),
                user.getRole().toString(),
                user.getEmail().getValue()
        );

        recordEvent(newEvent);
    }

    /**
     * Method that changes the email of the given user. Produces event that user changed the email
     *
     * @param user existing user entity in a database
     * @param newEmail new email for entity
     */
    public void changeEmail(User user, UserEmail newEmail) {
        UserEmail oldEmail = user.getEmail();
        user.setEmail(newEmail);
        user.setUpdatedAt(Instant.now());

        UserChangedEmailEvent newEvent = new UserChangedEmailEvent(
                user.getId(),
                oldEmail.getValue(),
                newEmail.getValue(),
                user.getRole().toString(),
                user.getName().getValue()
        );

        recordEvent(newEvent);
    }
}
